From 47dad3581884da17c6824fbe65ee7445951fb812 Mon Sep 17 00:00:00 2001
From: Pradeep M <pradeep.m@intrinsyc.com>
Date: Wed, 21 Nov 2018 14:54:47 +0530
Subject: [PATCH 301/334] [7389] add auto focus support for ov5640, set voltage
 levels of ldo's

*) set CSI0_AVDD & CSI1_AVD (l17 & l18 ldo's) to 2.8v as per the specs, needed for CAM0 & CAM1 slots
*) turn on ldo 'l23' for actuator
*) add auto focus support
*) set contintuous focus for 1080P & 720P resolutions (focus will trigger when there is scene change)
*) set single focus for 5MP resolution (focus will trigger and lens will pause/stay at current position)

Change-Id: I52fd913490e52cb92f09e9c13cc3d3cc5bc86df1
---
 arch/arm64/boot/dts/qcom/apq8096-openq.dtsi |   7 +
 drivers/media/i2c/ov5640.c                  | 429 +++++++++++++++++++-
 2 files changed, 434 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi b/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
index b8806a3bcb8e..1ffdccfebab9 100644
--- a/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
+++ b/arch/arm64/boot/dts/qcom/apq8096-openq.dtsi
@@ -105,10 +105,17 @@
 				};
 
 				l17 {
+					regulator-min-microvolt = <2800000>;
+					regulator-max-microvolt = <2800000>;
 					regulator-always-on;
 				};
 
 				l18 {
+					regulator-min-microvolt = <2800000>;
+					regulator-max-microvolt = <2800000>;
+					regulator-always-on;
+				};
+				l23 {
 					regulator-always-on;
 				};
 			};
diff --git a/drivers/media/i2c/ov5640.c b/drivers/media/i2c/ov5640.c
index eb0bcf30fac3..2ca11feb2e11 100644
--- a/drivers/media/i2c/ov5640.c
+++ b/drivers/media/i2c/ov5640.c
@@ -27,6 +27,7 @@
 #include <media/v4l2-device.h>
 #include <media/v4l2-fwnode.h>
 #include <media/v4l2-subdev.h>
+#include <linux/sched/signal.h>
 
 /* min/typical/max system clock (xclk) frequencies */
 #define OV5640_XCLK_MIN  6000000
@@ -209,6 +210,7 @@ struct ov5640_dev {
 	struct i2c_client *i2c_client;
 	struct v4l2_subdev sd;
 	struct media_pad pad;
+	struct device *dev;
 	struct v4l2_fwnode_endpoint ep; /* the parsed DT endpoint info */
 	struct clk *xclk; /* system clock to OV5640 */
 	u32 xclk_freq;
@@ -388,7 +390,271 @@ static const struct reg_value ov5640_setting_15fps_VGA_640_480[] = {
 	{0x4407, 0x04, 0, 0}, {0x460b, 0x35, 0, 0}, {0x460c, 0x22, 0, 0},
 	{0x3824, 0x02, 0, 0}, {0x5001, 0xa3, 0, 0},
 };
-
+static u8 autofocus_Config[] =
+{
+	0x02,0x0f,0x3a,0x02,0x08,0x61,0xc2,0x01,0x22,0x22,0x00,0x02,0x0f,0x08,0x30,
+	0x01,0x03,0x02,0x02,0xa6,0x30,0x02,0x03,0x02,0x02,0xa6,0x90,0x51,0xa5,0xe0,
+	0x78,0x93,0xf6,0xa3,0xe0,0x08,0xf6,0xa3,0xe0,0x08,0xf6,0xe5,0x1f,0x70,0x4f,
+	0x75,0x1e,0x20,0xd2,0x35,0xd3,0x78,0x4f,0xe6,0x94,0x00,0x18,0xe6,0x94,0x00,
+	0x40,0x07,0xe6,0xfe,0x08,0xe6,0xff,0x80,0x03,0x12,0x0b,0xb6,0x78,0x7e,0xa6,
+	0x06,0x08,0xa6,0x07,0x78,0x8b,0xa6,0x09,0x18,0x76,0x01,0x12,0x0b,0xb6,0x78,
+	0x4e,0xa6,0x06,0x08,0xa6,0x07,0x78,0x8b,0xe6,0x78,0x6e,0xf6,0x75,0x1f,0x01,
+	0x78,0x93,0xe6,0x78,0x90,0xf6,0x78,0x94,0xe6,0x78,0x91,0xf6,0x78,0x95,0xe6,
+	0x78,0x92,0xf6,0x22,0x79,0x90,0xe7,0xd3,0x78,0x93,0x96,0x40,0x05,0xe7,0x96,
+	0xff,0x80,0x08,0xc3,0x79,0x93,0xe7,0x78,0x90,0x96,0xff,0x78,0x88,0x76,0x00,
+	0x08,0xa6,0x07,0x79,0x91,0xe7,0xd3,0x78,0x94,0x96,0x40,0x05,0xe7,0x96,0xff,
+	0x80,0x08,0xc3,0x79,0x94,0xe7,0x78,0x91,0x96,0xff,0x12,0x0b,0xff,0x79,0x92,
+	0xe7,0xd3,0x78,0x95,0x96,0x40,0x05,0xe7,0x96,0xff,0x80,0x08,0xc3,0x79,0x95,
+	0xe7,0x78,0x92,0x96,0xff,0x12,0x0b,0xff,0x12,0x0b,0xb6,0x78,0x8a,0xe6,0x25,
+	0xe0,0x24,0x4e,0xf8,0xa6,0x06,0x08,0xa6,0x07,0x78,0x8a,0xe6,0x24,0x6e,0xf8,
+	0xa6,0x09,0x90,0x0e,0x93,0xe4,0x93,0x24,0xff,0xff,0xe4,0x34,0xff,0xfe,0x78,
+	0x8a,0xe6,0x24,0x01,0xfd,0xe4,0x33,0xfc,0xd3,0xed,0x9f,0xee,0x64,0x80,0xf8,
+	0xec,0x64,0x80,0x98,0x40,0x04,0x7f,0x00,0x80,0x05,0x78,0x8a,0xe6,0x04,0xff,
+	0x78,0x8a,0xa6,0x07,0xe5,0x1f,0xb4,0x01,0x0a,0xe6,0x60,0x03,0x02,0x02,0xa6,
+	0x75,0x1f,0x02,0x22,0x78,0x4e,0xe6,0xfe,0x08,0xe6,0xff,0x78,0x80,0xa6,0x06,
+	0x08,0xa6,0x07,0x78,0x4e,0xe6,0xfe,0x08,0xe6,0xff,0x78,0x82,0xa6,0x06,0x08,
+	0xa6,0x07,0x78,0x6e,0xe6,0x78,0x8c,0xf6,0x78,0x6e,0xe6,0x78,0x8d,0xf6,0x7f,
+	0x01,0x90,0x0e,0x93,0xe4,0x93,0xfe,0xef,0xc3,0x9e,0x50,0x5f,0xef,0x25,0xe0,
+	0x24,0x4f,0xf9,0xc3,0x78,0x81,0xe6,0x97,0x18,0xe6,0x19,0x97,0x50,0x0a,0x12,
+	0x0b,0xe7,0x78,0x80,0xa6,0x04,0x08,0xa6,0x05,0x74,0x6e,0x2f,0xf9,0x78,0x8c,
+	0xe6,0xc3,0x97,0x50,0x08,0x74,0x6e,0x2f,0xf8,0xe6,0x78,0x8c,0xf6,0xef,0x25,
+	0xe0,0x24,0x4f,0xf9,0xd3,0x78,0x83,0xe6,0x97,0x18,0xe6,0x19,0x97,0x40,0x0a,
+	0x12,0x0b,0xe7,0x78,0x82,0xa6,0x04,0x08,0xa6,0x05,0x74,0x6e,0x2f,0xf9,0x78,
+	0x8d,0xe6,0xd3,0x97,0x40,0x08,0x74,0x6e,0x2f,0xf8,0xe6,0x78,0x8d,0xf6,0x0f,
+	0x80,0x96,0xc3,0x79,0x81,0xe7,0x78,0x83,0x96,0xff,0x19,0xe7,0x18,0x96,0x78,
+	0x84,0xf6,0x08,0xa6,0x07,0xc3,0x79,0x8c,0xe7,0x78,0x8d,0x96,0x08,0xf6,0x12,
+	0x0b,0xf3,0x40,0x05,0x09,0xe7,0x08,0x80,0x06,0xc3,0x79,0x7f,0xe7,0x78,0x81,
+	0x96,0xff,0x19,0xe7,0x18,0x96,0xfe,0x78,0x86,0xa6,0x06,0x08,0xa6,0x07,0x79,
+	0x8c,0xe7,0xd3,0x78,0x8b,0x96,0x40,0x05,0xe7,0x96,0xff,0x80,0x08,0xc3,0x79,
+	0x8b,0xe7,0x78,0x8c,0x96,0xff,0x78,0x8f,0xa6,0x07,0xe5,0x1f,0x64,0x02,0x60,
+	0x03,0x02,0x02,0x92,0x90,0x0e,0x91,0x93,0xff,0x18,0xe6,0xc3,0x9f,0x40,0x03,
+	0x02,0x02,0xa6,0x78,0x84,0x12,0x0b,0xd8,0x12,0x0b,0xad,0x90,0x0e,0x8e,0x12,
+	0x0b,0xc6,0x78,0x80,0xe6,0xfe,0x08,0xe6,0xff,0x12,0x0c,0x09,0x7b,0x04,0x12,
+	0x0b,0x9b,0xc3,0x12,0x07,0xcf,0x50,0x64,0x90,0x0e,0x92,0xe4,0x93,0xff,0x78,
+	0x8f,0xe6,0x9f,0x40,0x02,0x80,0x11,0x90,0x0e,0x90,0xe4,0x93,0xff,0xd3,0x78,
+	0x89,0xe6,0x9f,0x18,0xe6,0x94,0x00,0x40,0x03,0x75,0x1f,0x05,0x78,0x86,0x12,
+	0x0b,0xd8,0x12,0x0b,0xad,0x90,0x0e,0x8f,0x12,0x0b,0xc6,0x12,0x0b,0xf3,0x40,
+	0x02,0x80,0x02,0x78,0x80,0xe6,0xfe,0x08,0xe6,0xff,0x12,0x0c,0x09,0x7b,0x10,
+	0x12,0x0b,0x9b,0xd3,0x12,0x07,0xcf,0x40,0x18,0x75,0x1f,0x05,0x22,0xe5,0x1f,
+	0xb4,0x05,0x0f,0xd2,0x01,0xc2,0x02,0xe4,0xf5,0x1f,0xf5,0x1e,0xd2,0x35,0xd2,
+	0x33,0xd2,0x36,0x22,0xe5,0x1f,0x70,0x72,0xf5,0x1e,0xd2,0x35,0xff,0xef,0x25,
+	0xe0,0x24,0x4e,0xf8,0xe4,0xf6,0x08,0xf6,0x0f,0xbf,0x34,0xf2,0x90,0x0e,0x94,
+	0xe4,0x93,0xff,0xe5,0x4b,0xc3,0x9f,0x50,0x04,0x7f,0x05,0x80,0x02,0x7f,0xfb,
+	0x78,0xbd,0xa6,0x07,0x12,0x0d,0x0e,0x40,0x04,0x7f,0x03,0x80,0x02,0x7f,0x30,
+	0x78,0xbc,0xa6,0x07,0xe6,0x18,0xf6,0x08,0xe6,0x78,0xb9,0xf6,0x78,0xbc,0xe6,
+	0x78,0xba,0xf6,0x78,0xbf,0x76,0x33,0xe4,0x08,0xf6,0x78,0xb8,0x76,0x01,0x75,
+	0x4a,0x02,0x78,0xb6,0xf6,0x08,0xf6,0x74,0xff,0x78,0xc1,0xf6,0x08,0xf6,0x75,
+	0x1f,0x01,0x78,0xbc,0xe6,0x75,0xf0,0x05,0xa4,0xf5,0x4b,0x12,0x09,0xdd,0xc2,
+	0x37,0x22,0x78,0xb8,0xe6,0xd3,0x94,0x00,0x40,0x02,0x16,0x22,0xe5,0x1f,0xb4,
+	0x05,0x23,0xe4,0xf5,0x1f,0xc2,0x01,0x78,0xb6,0xe6,0xfe,0x08,0xe6,0xff,0x78,
+	0x4e,0xa6,0x06,0x08,0xa6,0x07,0xa2,0x37,0xe4,0x33,0xf5,0x3c,0x90,0x30,0x28,
+	0xf0,0x75,0x1e,0x10,0xd2,0x35,0x22,0xe5,0x4b,0x75,0xf0,0x05,0x84,0x78,0xbc,
+	0xf6,0x90,0x0e,0x8c,0xe4,0x93,0xff,0x25,0xe0,0x24,0x0a,0xf8,0xe6,0xfc,0x08,
+	0xe6,0xfd,0x78,0xbc,0xe6,0x25,0xe0,0x24,0x4e,0xf8,0xa6,0x04,0x08,0xa6,0x05,
+	0xef,0x12,0x0d,0x52,0xd3,0x78,0xb7,0x96,0xee,0x18,0x96,0x40,0x0d,0x78,0xbc,
+	0xe6,0x78,0xb9,0xf6,0x78,0xb6,0xa6,0x06,0x08,0xa6,0x07,0x90,0x0e,0x8c,0xe4,
+	0x93,0x12,0x0d,0x52,0xc3,0x78,0xc2,0x96,0xee,0x18,0x96,0x50,0x0d,0x78,0xbc,
+	0xe6,0x78,0xba,0xf6,0x78,0xc1,0xa6,0x06,0x08,0xa6,0x07,0x78,0xb6,0xe6,0xfe,
+	0x08,0xe6,0xc3,0x78,0xc2,0x96,0xff,0xee,0x18,0x96,0x78,0xc3,0xf6,0x08,0xa6,
+	0x07,0x90,0x0e,0x96,0xe4,0x18,0x12,0x0d,0x15,0x40,0x02,0xd2,0x37,0x78,0xbc,
+	0xe6,0x08,0x26,0x08,0xf6,0xe5,0x1f,0x64,0x01,0x70,0x7a,0xe6,0xc3,0x78,0xc0,
+	0x12,0x0c,0xfe,0x40,0x08,0x12,0x0c,0xf9,0x50,0x03,0x02,0x04,0xf1,0x12,0x0d,
+	0x0c,0x40,0x04,0x7f,0xfe,0x80,0x02,0x7f,0x02,0x78,0xbd,0xa6,0x07,0x78,0xb9,
+	0xe6,0x24,0x03,0x78,0xbf,0xf6,0x78,0xb9,0xe6,0x24,0xfd,0x78,0xc0,0xf6,0x18,
+	0x12,0x0d,0x0e,0x40,0x04,0xe6,0xff,0x80,0x02,0x7f,0x00,0x12,0x0d,0x2e,0x40,
+	0x04,0xe6,0xff,0x80,0x02,0x7f,0x00,0x12,0x0d,0x3a,0x50,0x04,0xe6,0xff,0x80,
+	0x02,0x7f,0x33,0x12,0x0d,0x46,0x50,0x04,0xe6,0xff,0x80,0x02,0x7f,0x33,0x12,
+	0x0d,0x08,0x40,0x06,0x78,0xc0,0xe6,0xff,0x80,0x04,0x78,0xbf,0xe6,0xff,0x78,
+	0xbe,0xa6,0x07,0x75,0x1f,0x02,0x78,0xb8,0x76,0x01,0x02,0x04,0xf1,0xe5,0x1f,
+	0x64,0x02,0x70,0x77,0x78,0xbe,0xe6,0xff,0xc3,0x78,0xc0,0x12,0x0c,0xff,0x40,
+	0x05,0x12,0x0c,0xf9,0x40,0x64,0x12,0x0d,0x0c,0x40,0x04,0x7f,0xff,0x80,0x02,
+	0x7f,0x01,0x78,0xbd,0xa6,0x07,0x78,0xb9,0xe6,0x04,0x78,0xbf,0xf6,0x78,0xb9,
+	0xe6,0x14,0x78,0xc0,0xf6,0x18,0x12,0x0d,0x33,0x40,0x04,0xe6,0xff,0x80,0x02,
+	0x7f,0x00,0x12,0x0d,0x2e,0x40,0x04,0xe6,0xff,0x80,0x02,0x7f,0x00,0x12,0x0d,
+	0x3a,0x50,0x04,0xe6,0xff,0x80,0x02,0x7f,0x33,0x12,0x0d,0x46,0x50,0x04,0xe6,
+	0xff,0x80,0x02,0x7f,0x33,0x12,0x0d,0x08,0x40,0x06,0x78,0xc0,0xe6,0xff,0x80,
+	0x04,0x78,0xbf,0xe6,0xff,0x78,0xbe,0xa6,0x07,0x75,0x1f,0x03,0x78,0xb8,0x76,
+	0x01,0x80,0x20,0xe5,0x1f,0x64,0x03,0x70,0x26,0x78,0xbe,0xe6,0xff,0xc3,0x78,
+	0xc0,0x12,0x0c,0xff,0x40,0x05,0x12,0x0c,0xf9,0x40,0x09,0x78,0xb9,0xe6,0x78,
+	0xbe,0xf6,0x75,0x1f,0x04,0x78,0xbe,0xe6,0x75,0xf0,0x05,0xa4,0xf5,0x4b,0x02,
+	0x09,0xdd,0xe5,0x1f,0x64,0x04,0x70,0x25,0x90,0x0e,0x95,0x78,0xc3,0x12,0x0d,
+	0x15,0x40,0x04,0xd2,0x37,0x80,0x14,0x90,0x0e,0x97,0xe4,0x93,0xff,0x60,0x0c,
+	0x90,0x0e,0x77,0xe4,0x93,0xf5,0x4a,0x8f,0x4b,0x12,0x09,0xdd,0x75,0x1f,0x05,
+	0x22,0x90,0x38,0x04,0x78,0x52,0x12,0x0a,0xf1,0x90,0x38,0x00,0xe0,0xfe,0xa3,
+	0xe0,0xfd,0xed,0xff,0xc3,0x12,0x0a,0x92,0x90,0x38,0x10,0x12,0x0a,0x86,0x90,
+	0x38,0x06,0x78,0x54,0x12,0x0a,0xf1,0x90,0x38,0x02,0xe0,0xfe,0xa3,0xe0,0xfd,
+	0xed,0xff,0xc3,0x12,0x0a,0x92,0x90,0x38,0x12,0x12,0x0a,0x86,0xa3,0xe0,0xb4,
+	0x31,0x07,0x78,0x52,0x79,0x52,0x12,0x0b,0x11,0x90,0x38,0x14,0xe0,0xb4,0x71,
+	0x15,0x78,0x52,0xe6,0xfe,0x08,0xe6,0x78,0x02,0xce,0xc3,0x13,0xce,0x13,0xd8,
+	0xf9,0x79,0x53,0xf7,0xee,0x19,0xf7,0x90,0x38,0x15,0xe0,0xb4,0x31,0x07,0x78,
+	0x54,0x79,0x54,0x12,0x0b,0x11,0x90,0x38,0x15,0xe0,0xb4,0x71,0x15,0x78,0x54,
+	0xe6,0xfe,0x08,0xe6,0x78,0x02,0xce,0xc3,0x13,0xce,0x13,0xd8,0xf9,0x79,0x55,
+	0xf7,0xee,0x19,0xf7,0x79,0x52,0x12,0x0a,0xcd,0x09,0x12,0x0a,0xcd,0xaf,0x47,
+	0x12,0x0a,0xa6,0xe5,0x44,0xfb,0x7a,0x00,0xfd,0x7c,0x00,0x12,0x07,0x2b,0x78,
+	0x5a,0xa6,0x06,0x08,0xa6,0x07,0xaf,0x45,0x12,0x0a,0xa6,0xad,0x03,0x7c,0x00,
+	0x12,0x07,0x2b,0x78,0x56,0xa6,0x06,0x08,0xa6,0x07,0xaf,0x48,0x78,0x54,0x12,
+	0x0a,0xa8,0xe5,0x43,0xfb,0xfd,0x7c,0x00,0x12,0x07,0x2b,0x78,0x5c,0xa6,0x06,
+	0x08,0xa6,0x07,0xaf,0x46,0x7e,0x00,0x78,0x54,0x12,0x0a,0xaa,0xad,0x03,0x7c,
+	0x00,0x12,0x07,0x2b,0x78,0x58,0xa6,0x06,0x08,0xa6,0x07,0xc3,0x78,0x5b,0xe6,
+	0x94,0x08,0x18,0xe6,0x94,0x00,0x50,0x05,0x76,0x00,0x08,0x76,0x08,0xc3,0x78,
+	0x5d,0xe6,0x94,0x08,0x18,0xe6,0x94,0x00,0x50,0x05,0x76,0x00,0x08,0x76,0x08,
+	0x78,0x5a,0x12,0x0a,0xba,0xff,0xd3,0x78,0x57,0xe6,0x9f,0x18,0xe6,0x9e,0x40,
+	0x0e,0x78,0x5a,0xe6,0x13,0xfe,0x08,0xe6,0x78,0x57,0x12,0x0a,0xfc,0x80,0x04,
+	0x7e,0x00,0x7f,0x00,0x78,0x5e,0x12,0x0a,0xb2,0xff,0xd3,0x78,0x59,0xe6,0x9f,
+	0x18,0xe6,0x9e,0x40,0x0e,0x78,0x5c,0xe6,0x13,0xfe,0x08,0xe6,0x78,0x59,0x12,
+	0x0a,0xfc,0x80,0x04,0x7e,0x00,0x7f,0x00,0xe4,0xfc,0xfd,0x78,0x62,0x12,0x08,
+	0x23,0x78,0x5a,0x12,0x0a,0xba,0x78,0x57,0x26,0xff,0xee,0x18,0x36,0xfe,0x78,
+	0x66,0x12,0x0a,0xb2,0x78,0x59,0x26,0xff,0xee,0x18,0x36,0xfe,0xe4,0xfc,0xfd,
+	0x78,0x6a,0x12,0x08,0x23,0x12,0x0a,0xc2,0x78,0x66,0x12,0x08,0x16,0xd3,0x12,
+	0x07,0xcf,0x40,0x08,0x12,0x0a,0xc2,0x78,0x66,0x12,0x08,0x23,0x78,0x54,0x12,
+	0x0a,0xc4,0x78,0x6a,0x12,0x08,0x16,0xd3,0x12,0x07,0xcf,0x40,0x0a,0x78,0x54,
+	0x12,0x0a,0xc4,0x78,0x6a,0x12,0x08,0x23,0x78,0x61,0xe6,0x90,0x60,0x01,0xf0,
+	0x78,0x65,0xe6,0xa3,0xf0,0x78,0x69,0xe6,0xa3,0xf0,0x78,0x55,0xe6,0xa3,0xf0,
+	0x7d,0x01,0x78,0x61,0x12,0x0a,0xdd,0x24,0x01,0x12,0x0a,0x9a,0x78,0x65,0x12,
+	0x0a,0xdd,0x24,0x02,0x12,0x0a,0x9a,0x78,0x69,0x12,0x0a,0xdd,0x24,0x03,0x12,
+	0x0a,0x9a,0x78,0x6d,0x12,0x0a,0xdd,0x24,0x04,0x12,0x0a,0x9a,0x0d,0xbd,0x05,
+	0xd4,0x22,0xef,0x8d,0xf0,0xa4,0xa8,0xf0,0xcf,0x8c,0xf0,0xa4,0x28,0xce,0x8d,
+	0xf0,0xa4,0x2e,0xfe,0x22,0xbc,0x00,0x0b,0xbe,0x00,0x29,0xef,0x8d,0xf0,0x84,
+	0xff,0xad,0xf0,0x22,0xe4,0xcc,0xf8,0x75,0xf0,0x08,0xef,0x2f,0xff,0xee,0x33,
+	0xfe,0xec,0x33,0xfc,0xee,0x9d,0xec,0x98,0x40,0x05,0xfc,0xee,0x9d,0xfe,0x0f,
+	0xd5,0xf0,0xe9,0xe4,0xce,0xfd,0x22,0xed,0xf8,0xf5,0xf0,0xee,0x84,0x20,0xd2,
+	0x1c,0xfe,0xad,0xf0,0x75,0xf0,0x08,0xef,0x2f,0xff,0xed,0x33,0xfd,0x40,0x07,
+	0x98,0x50,0x06,0xd5,0xf0,0xf2,0x22,0xc3,0x98,0xfd,0x0f,0xd5,0xf0,0xea,0x22,
+	0xe8,0x8f,0xf0,0xa4,0xcc,0x8b,0xf0,0xa4,0x2c,0xfc,0xe9,0x8e,0xf0,0xa4,0x2c,
+	0xfc,0x8a,0xf0,0xed,0xa4,0x2c,0xfc,0xea,0x8e,0xf0,0xa4,0xcd,0xa8,0xf0,0x8b,
+	0xf0,0xa4,0x2d,0xcc,0x38,0x25,0xf0,0xfd,0xe9,0x8f,0xf0,0xa4,0x2c,0xcd,0x35,
+	0xf0,0xfc,0xeb,0x8e,0xf0,0xa4,0xfe,0xa9,0xf0,0xeb,0x8f,0xf0,0xa4,0xcf,0xc5,
+	0xf0,0x2e,0xcd,0x39,0xfe,0xe4,0x3c,0xfc,0xea,0xa4,0x2d,0xce,0x35,0xf0,0xfd,
+	0xe4,0x3c,0xfc,0x22,0xeb,0x9f,0xf5,0xf0,0xea,0x9e,0x42,0xf0,0xe9,0x9d,0x42,
+	0xf0,0xe8,0x9c,0x45,0xf0,0x22,0xe8,0x60,0x0f,0xec,0xc3,0x13,0xfc,0xed,0x13,
+	0xfd,0xee,0x13,0xfe,0xef,0x13,0xff,0xd8,0xf1,0x22,0xe8,0x60,0x0f,0xef,0xc3,
+	0x33,0xff,0xee,0x33,0xfe,0xed,0x33,0xfd,0xec,0x33,0xfc,0xd8,0xf1,0x22,0xe4,
+	0x93,0xfc,0x74,0x01,0x93,0xfd,0x74,0x02,0x93,0xfe,0x74,0x03,0x93,0xff,0x22,
+	0xe6,0xfb,0x08,0xe6,0xf9,0x08,0xe6,0xfa,0x08,0xe6,0xcb,0xf8,0x22,0xec,0xf6,
+	0x08,0xed,0xf6,0x08,0xee,0xf6,0x08,0xef,0xf6,0x22,0xa4,0x25,0x82,0xf5,0x82,
+	0xe5,0xf0,0x35,0x83,0xf5,0x83,0x22,0xd0,0x83,0xd0,0x82,0xf8,0xe4,0x93,0x70,
+	0x12,0x74,0x01,0x93,0x70,0x0d,0xa3,0xa3,0x93,0xf8,0x74,0x01,0x93,0xf5,0x82,
+	0x88,0x83,0xe4,0x73,0x74,0x02,0x93,0x68,0x60,0xef,0xa3,0xa3,0xa3,0x80,0xdf,
+	0xc0,0xe0,0xc0,0x83,0xc0,0x82,0xc0,0xd0,0x90,0x3f,0x0c,0xe0,0xf5,0x32,0xe5,
+	0x32,0x30,0xe3,0x74,0x30,0x36,0x66,0x90,0x60,0x19,0xe0,0xf5,0x0a,0xa3,0xe0,
+	0xf5,0x0b,0x90,0x60,0x1d,0xe0,0xf5,0x14,0xa3,0xe0,0xf5,0x15,0x90,0x60,0x21,
+	0xe0,0xf5,0x0c,0xa3,0xe0,0xf5,0x0d,0x90,0x60,0x29,0xe0,0xf5,0x0e,0xa3,0xe0,
+	0xf5,0x0f,0x90,0x60,0x31,0xe0,0xf5,0x10,0xa3,0xe0,0xf5,0x11,0x90,0x60,0x39,
+	0xe0,0xf5,0x12,0xa3,0xe0,0xf5,0x13,0x30,0x01,0x06,0x30,0x33,0x03,0xd3,0x80,
+	0x01,0xc3,0x92,0x09,0x30,0x02,0x06,0x30,0x33,0x03,0xd3,0x80,0x01,0xc3,0x92,
+	0x0a,0x30,0x33,0x0c,0x30,0x03,0x09,0x20,0x02,0x06,0x20,0x01,0x03,0xd3,0x80,
+	0x01,0xc3,0x92,0x0b,0x90,0x30,0x01,0xe0,0x44,0x40,0xf0,0xe0,0x54,0xbf,0xf0,
+	0xe5,0x32,0x30,0xe1,0x14,0x30,0x34,0x11,0x90,0x30,0x22,0xe0,0xf5,0x08,0xe4,
+	0xf0,0x30,0x00,0x03,0xd3,0x80,0x01,0xc3,0x92,0x08,0xe5,0x32,0x30,0xe5,0x12,
+	0x90,0x56,0xa1,0xe0,0xf5,0x09,0x30,0x31,0x09,0x30,0x05,0x03,0xd3,0x80,0x01,
+	0xc3,0x92,0x0d,0x90,0x3f,0x0c,0xe5,0x32,0xf0,0xd0,0xd0,0xd0,0x82,0xd0,0x83,
+	0xd0,0xe0,0x32,0x85,0x08,0x41,0x90,0x30,0x24,0xe0,0xf5,0x3d,0xa3,0xe0,0xf5,
+	0x3e,0xa3,0xe0,0xf5,0x3f,0xa3,0xe0,0xf5,0x40,0xa3,0xe0,0xf5,0x3c,0xd2,0x34,
+	0xe5,0x41,0x12,0x08,0x3b,0x09,0x6c,0x03,0x09,0x79,0x04,0x09,0x8a,0x05,0x09,
+	0x8d,0x06,0x09,0x96,0x08,0x09,0xa6,0x12,0x09,0xab,0x1a,0x09,0xb6,0x1b,0x09,
+	0xa6,0x80,0x09,0xa6,0x81,0x09,0xbe,0xec,0x00,0x00,0x09,0xdc,0x12,0x0f,0x2c,
+	0xd2,0x36,0xd2,0x01,0xc2,0x02,0x12,0x0f,0x31,0x22,0xd2,0x33,0xd2,0x36,0xe5,
+	0x3d,0xd3,0x94,0x00,0x40,0x03,0x12,0x0f,0x2c,0xd2,0x03,0x22,0xd2,0x03,0x22,
+	0xc2,0x03,0x20,0x01,0x4a,0x30,0x02,0x2c,0x22,0xc2,0x01,0xc2,0x02,0xc2,0x03,
+	0x12,0x0c,0x88,0x75,0x1e,0x70,0xd2,0x35,0x80,0x1b,0x12,0x0b,0x1b,0x80,0x16,
+	0x85,0x40,0x4a,0x85,0x3c,0x4b,0x12,0x09,0xdd,0x80,0x0b,0x85,0x4a,0x40,0x85,
+	0x4b,0x3c,0x80,0x03,0x12,0x0d,0x5d,0x90,0x30,0x24,0xe5,0x3d,0xf0,0xa3,0xe5,
+	0x3e,0xf0,0xa3,0xe5,0x3f,0xf0,0xa3,0xe5,0x40,0xf0,0xa3,0xe5,0x3c,0xf0,0x90,
+	0x30,0x23,0xe4,0xf0,0x22,0x12,0x0e,0xdf,0xc3,0x90,0x0e,0x7c,0x74,0x01,0x93,
+	0x9f,0xff,0xe4,0x93,0x9e,0xfe,0xe4,0x8f,0x3b,0x8e,0x3a,0xf5,0x39,0xf5,0x38,
+	0xab,0x3b,0xaa,0x3a,0xa9,0x39,0xa8,0x38,0xaf,0x4b,0xfc,0xfd,0xfe,0x12,0x07,
+	0x80,0x12,0x0e,0xca,0x78,0x08,0x12,0x07,0xe0,0x12,0x0e,0xca,0x12,0x0e,0xdf,
+	0xe4,0xfc,0xfd,0xe5,0x3b,0x2f,0xf5,0x3b,0xe5,0x3a,0x3e,0xf5,0x3a,0xed,0x35,
+	0x39,0xf5,0x39,0xec,0x35,0x38,0xf5,0x38,0xaf,0x3b,0xae,0x3a,0xad,0x39,0xfc,
+	0x90,0x0e,0x69,0xe4,0x12,0x0e,0xea,0x12,0x0e,0xca,0xe4,0x85,0x4a,0x37,0xf5,
+	0x36,0xf5,0x35,0xf5,0x34,0xaf,0x37,0xae,0x36,0xad,0x35,0xac,0x34,0xa3,0x12,
+	0x0e,0xea,0x8f,0x37,0x8e,0x36,0x8d,0x35,0x8c,0x34,0xe5,0x3b,0x45,0x37,0xf5,
+	0x3b,0xe5,0x3a,0x45,0x36,0xf5,0x3a,0xe5,0x39,0x45,0x35,0xf5,0x39,0xe5,0x38,
+	0x45,0x34,0xf5,0x38,0xe4,0xf5,0x22,0xf5,0x23,0x85,0x3b,0x31,0x85,0x3a,0x30,
+	0x85,0x39,0x2f,0x85,0x38,0x2e,0x02,0x0e,0x9c,0xe0,0xa3,0xe0,0x75,0xf0,0x02,
+	0xa4,0xff,0xae,0xf0,0xc3,0x08,0xe6,0x9f,0xf6,0x18,0xe6,0x9e,0xf6,0x22,0xff,
+	0xe5,0xf0,0x34,0x60,0x8f,0x82,0xf5,0x83,0xec,0xf0,0x22,0x78,0x52,0x7e,0x00,
+	0xe6,0xfc,0x08,0xe6,0xfd,0x02,0x07,0x19,0xe4,0xfc,0xfd,0x12,0x08,0x23,0x78,
+	0x5c,0xe6,0xc3,0x13,0xfe,0x08,0xe6,0x13,0x22,0x78,0x52,0xe6,0xfe,0x08,0xe6,
+	0xff,0xe4,0xfc,0xfd,0x22,0xe7,0xc4,0xf8,0x54,0xf0,0xc8,0x68,0xf7,0x09,0xe7,
+	0xc4,0x54,0x0f,0x48,0xf7,0x22,0xe6,0xfc,0xed,0x75,0xf0,0x04,0xa4,0x22,0x12,
+	0x08,0x06,0x8f,0x48,0x8e,0x47,0x8d,0x46,0x8c,0x45,0x22,0xe0,0xfe,0xa3,0xe0,
+	0xfd,0xee,0xf6,0xed,0x08,0xf6,0x22,0x13,0xff,0xc3,0xe6,0x9f,0xff,0x18,0xe6,
+	0x9e,0xfe,0x22,0xfb,0xd3,0xed,0x9b,0x74,0x80,0xf8,0x6c,0x98,0x22,0xe6,0xc3,
+	0x13,0xf7,0x08,0xe6,0x13,0x09,0xf7,0x22,0xe5,0x3c,0xd3,0x94,0x01,0x40,0x0b,
+	0x90,0x0e,0x88,0x12,0x0a,0xe5,0x90,0x0e,0x86,0x80,0x09,0x90,0x0e,0x82,0x12,
+	0x0a,0xe5,0x90,0x0e,0x80,0xe4,0x93,0xf5,0x44,0xa3,0xe4,0x93,0xf5,0x43,0xe5,
+	0x3c,0xd3,0x94,0x00,0x40,0x06,0x85,0x3d,0x45,0x85,0x3e,0x46,0xe5,0x47,0xc3,
+	0x13,0xff,0xe5,0x45,0xc3,0x9f,0x50,0x02,0x8f,0x45,0xe5,0x48,0xc3,0x13,0xff,
+	0xe5,0x46,0xc3,0x9f,0x50,0x02,0x8f,0x46,0xe5,0x47,0xc3,0x13,0xff,0xfd,0xe5,
+	0x45,0x2d,0xfd,0xe4,0x33,0xfc,0xe5,0x44,0x12,0x0b,0x07,0x40,0x05,0xe5,0x44,
+	0x9f,0xf5,0x45,0xe5,0x48,0xc3,0x13,0xff,0xfd,0xe5,0x46,0x2d,0xfd,0xe4,0x33,
+	0xfc,0xe5,0x43,0x12,0x0b,0x07,0x40,0x05,0xe5,0x43,0x9f,0xf5,0x46,0x02,0x05,
+	0x29,0xad,0x39,0xac,0x38,0xfa,0xf9,0xf8,0x12,0x07,0x80,0x8f,0x3b,0x8e,0x3a,
+	0x8d,0x39,0x8c,0x38,0xab,0x37,0xaa,0x36,0xa9,0x35,0xa8,0x34,0x22,0x90,0x0e,
+	0x8c,0xe4,0x93,0x25,0xe0,0x24,0x0a,0xf8,0xe6,0xfe,0x08,0xe6,0xff,0x22,0x93,
+	0xff,0xe4,0xfc,0xfd,0xfe,0x12,0x07,0x80,0x8f,0x37,0x8e,0x36,0x8d,0x35,0x8c,
+	0x34,0x22,0xe6,0xfe,0x08,0xe6,0xff,0xe4,0x8f,0x37,0x8e,0x36,0xf5,0x35,0xf5,
+	0x34,0x22,0xef,0x25,0xe0,0x24,0x4e,0xf8,0xe6,0xfc,0x08,0xe6,0xfd,0x22,0xd3,
+	0x79,0x81,0xe7,0x78,0x7f,0x96,0x19,0xe7,0x18,0x96,0x22,0x78,0x89,0xef,0x26,
+	0xf6,0x18,0xe4,0x36,0xf6,0x22,0xe4,0x8f,0x3b,0x8e,0x3a,0xf5,0x39,0xf5,0x38,
+	0x22,0x75,0x89,0x03,0x75,0xa8,0x01,0x75,0xb8,0x04,0x75,0x34,0xff,0x75,0x35,
+	0x0e,0x75,0x36,0x15,0x75,0x37,0x0d,0x12,0x0d,0xc0,0x12,0x00,0x09,0x12,0x0b,
+	0x1b,0x12,0x00,0x06,0xd2,0x00,0xd2,0x34,0xd2,0xaf,0x75,0x34,0xff,0x75,0x35,
+	0x0e,0x75,0x36,0x49,0x75,0x37,0x03,0x12,0x0d,0xc0,0x30,0x08,0x09,0xc2,0x34,
+	0x12,0x09,0x27,0xc2,0x08,0xd2,0x34,0x30,0x0b,0x09,0xc2,0x36,0x12,0x00,0x0e,
+	0xc2,0x0b,0xd2,0x36,0x30,0x09,0x09,0xc2,0x36,0x12,0x02,0xa7,0xc2,0x09,0xd2,
+	0x36,0x30,0x0e,0x03,0x12,0x05,0x29,0x30,0x35,0xd3,0x90,0x30,0x29,0xe5,0x1e,
+	0xf0,0xb4,0x10,0x05,0x90,0x30,0x23,0xe4,0xf0,0xc2,0x35,0x80,0xc1,0xe4,0xf5,
+	0x4b,0x90,0x0e,0x7a,0x93,0xff,0xe4,0x8f,0x37,0xf5,0x36,0xf5,0x35,0xf5,0x34,
+	0xaf,0x37,0xae,0x36,0xad,0x35,0xac,0x34,0x90,0x0e,0x6a,0x12,0x0e,0xea,0x8f,
+	0x37,0x8e,0x36,0x8d,0x35,0x8c,0x34,0x90,0x0e,0x72,0x12,0x08,0x06,0xef,0x45,
+	0x37,0xf5,0x37,0xee,0x45,0x36,0xf5,0x36,0xed,0x45,0x35,0xf5,0x35,0xec,0x45,
+	0x34,0xf5,0x34,0xe4,0xf5,0x22,0xf5,0x23,0x85,0x37,0x31,0x85,0x36,0x30,0x85,
+	0x35,0x2f,0x85,0x34,0x2e,0x12,0x0e,0x9c,0xe4,0xf5,0x22,0xf5,0x23,0x90,0x0e,
+	0x72,0x12,0x0e,0xd3,0x12,0x0e,0x9c,0xe4,0xf5,0x22,0xf5,0x23,0x90,0x0e,0x6e,
+	0x12,0x0e,0xd3,0x02,0x0e,0x9c,0x78,0xbe,0xe6,0xd3,0x08,0xff,0xe6,0x64,0x80,
+	0xf8,0xef,0x64,0x80,0x98,0x22,0x78,0xc0,0xa6,0x07,0x78,0xbd,0xd3,0xe6,0x64,
+	0x80,0x94,0x80,0x22,0x93,0xff,0x7e,0x00,0xe6,0xfc,0x08,0xe6,0xfd,0x12,0x07,
+	0x19,0x78,0xc1,0xe6,0xfc,0x08,0xe6,0xfd,0xd3,0xef,0x9d,0xee,0x9c,0x22,0x78,
+	0xbf,0xa6,0x07,0x08,0xd3,0xe6,0x64,0x80,0x94,0x80,0x22,0x78,0xc0,0xa6,0x07,
+	0xc3,0x18,0xe6,0x64,0x80,0x94,0xb3,0x22,0x78,0xbf,0xa6,0x07,0xc3,0x08,0xe6,
+	0x64,0x80,0x94,0xb3,0x22,0x25,0xe0,0x24,0x0a,0xf8,0xe6,0xfe,0x08,0xe6,0xff,
+	0x22,0xe5,0x40,0x24,0xf2,0xf5,0x37,0xe5,0x3f,0x34,0x43,0xf5,0x36,0xe5,0x3e,
+	0x34,0xa2,0xf5,0x35,0xe5,0x3d,0x34,0x28,0xf5,0x34,0xe5,0x37,0xff,0xe4,0xfe,
+	0xfd,0xfc,0x78,0x18,0x12,0x07,0xf3,0x8f,0x40,0x8e,0x3f,0x8d,0x3e,0x8c,0x3d,
+	0xe5,0x37,0x54,0xa0,0xff,0xe5,0x36,0xfe,0xe4,0xfd,0xfc,0x78,0x07,0x12,0x07,
+	0xe0,0x78,0x10,0x12,0x0e,0xf0,0xe4,0xff,0xfe,0xe5,0x35,0xfd,0xe4,0xfc,0x78,
+	0x0e,0x12,0x07,0xe0,0x12,0x0e,0xf3,0xe4,0xff,0xfe,0xfd,0xe5,0x34,0xfc,0x78,
+	0x18,0x12,0x07,0xe0,0x78,0x08,0x12,0x0e,0xf0,0x22,0xae,0x35,0xaf,0x36,0xe4,
+	0xfd,0xed,0xc3,0x95,0x37,0x50,0x33,0x12,0x0f,0x46,0xe4,0x93,0xf5,0x38,0x74,
+	0x01,0x93,0xf5,0x39,0x45,0x38,0x60,0x23,0x85,0x39,0x82,0x85,0x38,0x83,0xe0,
+	0xfc,0x12,0x0f,0x46,0x74,0x03,0x93,0x52,0x04,0x12,0x0f,0x46,0x74,0x02,0x93,
+	0x42,0x04,0x85,0x39,0x82,0x85,0x38,0x83,0xec,0xf0,0x0d,0x80,0xc7,0x22,0x13,
+	0x10,0x17,0x10,0x50,0x38,0x53,0x54,0x44,0x20,0x20,0x20,0x20,0x20,0x13,0x01,
+	0x10,0x01,0x56,0x40,0x1a,0x30,0x29,0x7e,0x00,0x30,0x04,0x20,0xdf,0x30,0x05,
+	0x40,0xbf,0x50,0x03,0x00,0xfd,0x50,0x27,0x01,0xfe,0x60,0x00,0x11,0x00,0x3f,
+	0x05,0x30,0x00,0x3f,0x06,0x22,0x00,0x3f,0x01,0x2a,0x00,0x3f,0x02,0x00,0x00,
+	0x36,0x06,0x07,0x00,0x3f,0x0b,0x0f,0xf0,0x00,0x00,0x00,0x00,0x30,0x01,0x40,
+	0xbf,0x30,0x01,0x00,0xbf,0x30,0x29,0x70,0x00,0x3a,0x00,0x00,0xff,0x3a,0x00,
+	0x00,0xff,0x36,0x03,0x36,0x02,0x41,0x44,0x58,0x20,0x18,0x10,0x0a,0x04,0x04,
+	0x00,0x03,0xff,0x64,0x00,0x00,0x80,0x00,0x00,0x00,0x00,0x00,0x00,0x02,0x04,
+	0x06,0x06,0x00,0x02,0x65,0x00,0x7a,0x50,0x3c,0x28,0x1e,0x10,0x10,0x50,0x2d,
+	0x28,0x16,0x10,0x10,0x02,0x00,0x10,0x30,0x0a,0x04,0x05,0x08,0x64,0x0a,0x05,
+	0x00,0x00,0xa5,0x5a,0x00,0xa2,0xaf,0x92,0x32,0xc2,0xaf,0xe5,0x23,0x45,0x22,
+	0x90,0x0e,0x5d,0x60,0x0e,0x12,0x0f,0x21,0xe0,0xf5,0x2c,0x12,0x0f,0x1e,0xe0,
+	0xf5,0x2d,0x80,0x0c,0x12,0x0f,0x21,0xe5,0x30,0xf0,0x12,0x0f,0x1e,0xe5,0x31,
+	0xf0,0xa2,0x32,0x92,0xaf,0x22,0x8f,0x3b,0x8e,0x3a,0x8d,0x39,0x8c,0x38,0x22,
+	0x12,0x08,0x06,0x8f,0x31,0x8e,0x30,0x8d,0x2f,0x8c,0x2e,0x22,0x90,0x0e,0x7e,
+	0xe4,0x93,0xfe,0x74,0x01,0x93,0xff,0x22,0x93,0xf9,0xf8,0x02,0x07,0xf3,0x12,
+	0x07,0xf3,0xe5,0x40,0x2f,0xf5,0x40,0xe5,0x3f,0x3e,0xf5,0x3f,0xe5,0x3e,0x3d,
+	0xf5,0x3e,0xe5,0x3d,0x3c,0xf5,0x3d,0x22,0xc0,0xe0,0xc0,0x83,0xc0,0x82,0x90,
+	0x3f,0x0d,0xe0,0xf5,0x33,0xe5,0x33,0xf0,0xd0,0x82,0xd0,0x83,0xd0,0xe0,0x32,
+	0x90,0x0e,0x5f,0xe4,0x93,0xfe,0x74,0x01,0x93,0xf5,0x82,0x8e,0x83,0x22,0xd2,
+	0x01,0xc2,0x02,0xe4,0xf5,0x1f,0xf5,0x1e,0xd2,0x35,0xd2,0x33,0x22,0x78,0x7f,
+	0xe4,0xf6,0xd8,0xfd,0x75,0x81,0xcd,0x02,0x0c,0x13,0x8f,0x82,0x8e,0x83,0x75,
+	0xf0,0x04,0xed,0x02,0x08,0x2f
+};
 static const struct reg_value ov5640_setting_30fps_XGA_1024_768[] = {
 
 	{0x3035, 0x14, 0, 0}, {0x3036, 0x38, 0, 0}, {0x3c07, 0x08, 0, 0},
@@ -1741,6 +2007,107 @@ static int ov5640_set_mode_direct(struct ov5640_dev *sensor,
 
 	return ov5640_load_regs(sensor, mode);
 }
+static int ov5640_af_init(struct ov5640_dev *ov5640)
+{
+	int totalCnt = 0;
+	int sentCnt = 0;
+	u16 addr = 0x8000;
+	int ret = 0;
+
+	ret = ov5640_write_reg(ov5640, 0x3000, 0x20); // reset mcu
+
+	if (ret < 0)
+	{
+		dev_err(ov5640->dev, "Focus initialization: cant reset the MCU\n");
+		return ret;
+	}
+
+	msleep(10);
+	totalCnt = ARRAY_SIZE(autofocus_Config);
+
+	for (sentCnt = 0; sentCnt < totalCnt; sentCnt++, addr++)
+	{
+		ret = ov5640_write_reg(ov5640, addr, autofocus_Config[sentCnt]);
+
+		if (ret < 0)
+		{
+			dev_err(ov5640->dev, "Focus initialization: failed writing reg: 0x%x\n", addr);
+			return ret;
+		}
+	}
+	return(ret);
+}
+static int ov5640_af_init1(struct ov5640_dev *ov5640)
+{
+	u8 state=0x8F;
+	int remaining = 0;
+	int ret = 0;
+
+	ret = ov5640_write_reg(ov5640, 0x3000, 0x20); // reset mcu
+	if (ret < 0)
+	{
+		dev_err(ov5640->dev, "Focus initialization: cant reset the MCU\n");
+		return ret;
+	}
+	msleep(10);
+
+	ov5640_write_reg(ov5640, 0x3022, 0x00);
+	ov5640_write_reg(ov5640, 0x3023, 0x00);
+	ov5640_write_reg(ov5640, 0x3024, 0x00);
+	ov5640_write_reg(ov5640, 0x3025, 0x00);
+	ov5640_write_reg(ov5640, 0x3026, 0x00);
+	ov5640_write_reg(ov5640, 0x3027, 0x00);
+	ov5640_write_reg(ov5640, 0x3028, 0x00);
+	ov5640_write_reg(ov5640, 0x3029, 0x7f);
+	ov5640_write_reg(ov5640, 0x3000, 0x00);
+
+	// Wait for up tp 1000ms to converge
+	for (remaining = 1000;;)
+	{
+		ret = (u8)ov5640_read_reg(ov5640, 0x3029, &state);
+
+		if (ret < 0)
+		{
+			dev_err(ov5640->dev, "Focus initialization: cant read status reg 0x3029\n");
+			break;
+		}
+
+		if (state == 0x70)
+		{
+			dev_err(ov5640->dev, "Focus initialization:  V4L2_AUTO_FOCUS_STATUS_IDLE\n");
+			break;
+		}
+
+		remaining -= (10 - msleep_interruptible(10));
+		if (signal_pending(current))
+			break;
+
+		if (remaining <= 0)
+		{
+			dev_err(ov5640->dev, "Focus initialization: timed out waiting to complete\n");
+			ret = -1;
+			break;
+		}
+	}
+
+	return(ret);
+}
+static int ov5640_setup_af(struct ov5640_dev *ov5640)
+{
+	int ret = 0;
+
+	dev_err(ov5640->dev,"Autofocus initialization started\n");
+	ret = ov5640_af_init(ov5640);
+
+	if (ret < 0) {
+		dev_err(ov5640->dev, "Autofocus initialization failed\n");
+		return(ret);
+	}
+
+	dev_err(ov5640->dev, "Autofocus initialization succeeded\n");
+
+	return(0);
+}
 
 static int ov5640_set_mode(struct ov5640_dev *sensor,
 			   const struct ov5640_mode_info *orig_mode)
@@ -1750,6 +2117,8 @@ static int ov5640_set_mode(struct ov5640_dev *sensor,
 	bool auto_gain = sensor->ctrls.auto_gain->val == 1;
 	bool auto_exp =  sensor->ctrls.auto_exp->val == V4L2_EXPOSURE_AUTO;
 	int ret;
+	u8 state = 0x8F;
+	int remaining ;
 
 	if (!orig_mode)
 		orig_mode = mode;
@@ -1768,6 +2137,13 @@ static int ov5640_set_mode(struct ov5640_dev *sensor,
 		if (ret)
 			goto restore_auto_gain;
 	}
+	/* focus initialization */
+	ret = ov5640_af_init1(sensor);
+	if (ret < 0) {
+		dev_err(sensor->dev, "%s: could not initialize mcu for focus control\n",
+			__func__);
+		return(ret);
+	}
 
 	if ((dn_mode == SUBSAMPLING && orig_dn_mode == SCALING) ||
 	    (dn_mode == SCALING && orig_dn_mode == SUBSAMPLING)) {
@@ -1807,6 +2183,48 @@ static int ov5640_set_mode(struct ov5640_dev *sensor,
 	if (ret < 0)
 		return ret;
 
+	/* single focus */
+	if(mode->hact == 2592) {
+		ov5640_write_reg(sensor, 0x3023, 0x01);
+		ov5640_write_reg(sensor, 0x3022,0x81);
+		mdelay(50);
+		ov5640_write_reg(sensor, 0x3022, 0x03);
+		// Wait for up to 7000ms to converge
+		for (remaining = 7000;;)
+		{
+			ret = (u8)ov5640_read_reg(sensor, 0x3029, &state);
+
+			if (ret < 0)
+			{
+				break;
+			}
+
+			if (state == 0x10)
+			{
+				remaining = 7000 - remaining;
+				dev_err(sensor->dev, "single Focus succeeded in %dms\n", remaining);
+				break;
+			}
+
+			remaining -= (30 - msleep_interruptible(30));
+			if (signal_pending(current))
+				break;
+
+			if (remaining <= 0)
+			{
+				dev_err(sensor->dev,"Focus failed: timed out\n");
+				ret = -1;
+				break;
+			}
+
+		}
+		ov5640_write_reg(sensor, 0x3022, 0x06);
+	} else {
+		/* continuous focus */
+		ov5640_write_reg(sensor, 0x3023,0x01);
+		ov5640_write_reg(sensor, 0x3022,0x04);
+	}
+
 	sensor->pending_mode_change = false;
 	sensor->last_mode = mode;
 
@@ -2788,12 +3206,13 @@ static int ov5640_probe(struct i2c_client *client,
 	struct v4l2_mbus_framefmt *fmt;
 	u32 rotation;
 	int ret;
-
+	u16 chip_id;
 	sensor = devm_kzalloc(dev, sizeof(*sensor), GFP_KERNEL);
 	if (!sensor)
 		return -ENOMEM;
 
 	sensor->i2c_client = client;
+	sensor->dev = dev;
 	fmt = &sensor->fmt;
 	fmt->code = ov5640_formats[0].code;
 	fmt->colorspace = ov5640_formats[0].colorspace;
@@ -2881,6 +3300,12 @@ static int ov5640_probe(struct i2c_client *client,
 
 	mutex_init(&sensor->lock);
 
+	ov5640_set_power_on(sensor);
+	ov5640_read_reg16(sensor, OV5640_REG_CHIP_ID, &chip_id);
+	pr_err(">>>>>>>>>camera sensor chipid>>>>> 0x%x \n",chip_id);
+	/*Load auto focus firmware*/
+	ov5640_setup_af(sensor);
+
 	ret = ov5640_init_controls(sensor);
 	if (ret)
 		goto entity_cleanup;
-- 
2.24.1

